catboost/private/libs/algo/greedy_tensor_search.cpp

```diff
@@ -1385,62 +1459,170 @@ static TNonSymmetricTreeStructure GreedyTensorSearchDepthwise(
         TVector<TIndexType> splittedLeafs;
         TVector<TIndexType> nextLevelLeafs;
         const size_t maxFeatureValueCount = CalcMaxFeatureValueCount(*fold, candidatesContexts);
-        for (TIndexType leafToSplit : curLevelLeafs) {
-            const auto& leafBounds = ctx->SampledDocs.LeavesBounds[leafToSplit];
+
+        const auto CalcBestScoreAndCandidate = [&data,&ctx,&scoreStDev,
+                                                &candidatesContexts, &fold,
+                                                &maxBucketCount,
+                                                &maxFeatureValueCount, &maxSplitEnsamples]
+                                                                      (TBucketStats* stats, TBucketStats* pstats,
+                                                                       TBucketStats* sstats, TIndexType id,
+                                                                       double& gainLocal,
+                                                                       const TCandidateInfo** bestSplitCandidateLocal,
+                                                                       TSplit& bestSplitLocal){
```

Comment:
1. all input parameters go first, and then all output parameters
2. all output parameters are passed by pointer

So this should be
```c++
[&] (
    TIndexType id,
    TBucketStats* stats,
    TBucketStats* pstats,
    TBucketStats* sstats,
    double* gainLocal,
    const TCandidateInfo** bestSplitCandidateLocal,
    TSplit* bestSplitLocal
) {
    ...
};
```

Or even better
```c++
[&] (
    TIndexType id,
    TArrayRef<TBucketStats> stats,
    TArrayRef<TBucketStats> pstats,
    TArrayRef<TBucketStats> sstats,
    double* gainLocal,
    TArrayRef<const TCandidateInfo*> bestSplitCandidateLocal,
    TSplit* bestSplitLocal
) {
    ...
};
```

Reply:
done